akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/EventsourcedStashReferenceManagement.scala

```diff
@@ -0,0 +1,35 @@
+/*
+ * Copyright (C) 2018 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.persistence.typed.internal
+
+import akka.actor.typed.scaladsl.{ ActorContext, Behaviors, StashBuffer }
+import akka.actor.typed.{ PostStop, Signal }
+import akka.annotation.InternalApi
+import akka.persistence.typed.internal.EventsourcedBehavior.InternalProtocol
+import akka.util.OptionVal
+
+@InternalApi
+trait EventsourcedStashReferenceManagement {
+
+  private var stashBuffer: OptionVal[StashBuffer[InternalProtocol]] = OptionVal.None
+
+  def stashBuffer(settings: EventsourcedSettings): StashBuffer[InternalProtocol] = {
+    val buffer: StashBuffer[InternalProtocol] = stashBuffer match {
+      case OptionVal.Some(value) ⇒ value
+      case _                     ⇒ StashBuffer(settings.stashCapacity)
+    }
+    this.stashBuffer = OptionVal.Some(buffer)
+    stashBuffer.get
+  }
+
+  def onSignalCleanup: (ActorContext[InternalProtocol], Signal) ⇒ Unit = {
+    case (ctx, PostStop) ⇒
+      stashBuffer match {
+        case OptionVal.Some(buffer) ⇒ buffer.unstashAll(ctx, Behaviors.ignore)
+        case _                      ⇒ Unit
+      }
+    case _ ⇒ Unit
+  }
+}
```

Comment:
I don't think this is right, It doesn't align with the general design of the new typed actor as it mixes in mutable state (like how it was done in the old persistent actor) into an object that is actually replaced with another behavior pretty quickly. I think the state should be completely moved into setup.

Reply:
The problem is that the stash must survive a restart. Thoughts about design here https://github.com/akka/akka/pull/24782#comment-177073511

I agree though that it is questionable. Let’s wait with merging this, perhaps someone has some better brilliant idea.

Reply:
Ah, I see, I missed that comment before. If we are to keep this solution I think there should be a comment explaining the why in the sources.

Reply:
If you are in control of the whole chain, Behaviors.setup -> Behaviors.supervise -> Behaviors.immutable, it's not a problem because then you can create and hold the StashBuffer to the left of the supervise. Here the user may add a supervise before our stuff.

Reply:
> Let’s wait with merging this, perhaps someone has some better brilliant idea.

One thing I can think of is to manage `StashBuffer` in separate actor/behavior (smth called like `StashReferenceBehavior`) and expand `InternalProtocol` with command like `ObtainStash`/`ClearStash`/`StashObtained(stash)` to make it all more transparent. However it seems that potential behavior reference (`StashReferenceBehavior` here) would have to be managed separately which brings us to basically the same solution which is implemented now.

Fork and branch are going nowhere and in case anybody comes up with something better we could redesign solution in this PR but as of now it at least fixes pending issue with restart of persistent actor.

Reply:
@johanandren any more thoughts about this?